### ISSUE TITLE
document.documentElement === null causes TypeError during loading of react

### ISSUE THREAD

**Issue by rsms** [facebook/react#7074]
[bug]

**Behavior**

To reproduce:
- Load react before the HTML document has loaded.

In my case I load react into a Chromium web view (CEF 3.2743.1432.g66425de_macosx64) just after the "window" object is ready. There's a natural race condition here where the HTML document might or might not have been parsed yet, meaning that `document.documentElement === null` in some cases.

When `document.documentElement` is null, the test `('textContent' in document.documentElement)` in [src/renderers/dom/client/utils/setTextContent.js](https://github.com/facebook/react/blob/40f6d3eacad563c67bea0aa1c2220de6239ece43/src/renderers/dom/client/utils/setTextContent.js#L41) causes an exception to be thrown:

``` js
Uncaught TypeError: Cannot use 'in' operator to search for 'textContent' in null
```

**What is the expected behavior?**

That react gets loaded without errors independently of the HTML document load state. I expect to control when the HTML document is accessed (i.e. not implicitly by React during module loading.)

_Possible fix:_ The code in [setTextContent.js](https://github.com/facebook/react/blob/40f6d3eacad563c67bea0aa1c2220de6239ece43/src/renderers/dom/client/utils/setTextContent.js#L41) should check `document.documentElement` before attempting to perform `in` on the object.

**Which versions of React, and which browser / OS are affected by this issue? Did this work in previous versions of React?**
- macOS 10.11.5 (15F34)
- Chromium Embedded Framework 3.2743.1432.g66425de_macosx64
- https://fb.me/react-15.1.0.min.js (also confirmed with https://fb.me/react-15.0.2.min.js)

**Comment by Aweary** [comment]
How are you loading React before the HTML document? I would imagine that not having access to the `documentElement` would [cause other issues](https://github.com/facebook/react/search?utf8=%E2%9C%93&q=documentElement) outside of `setTextContent`

**Comment by rsms** [comment]
@Aweary There are at least two ways I've been able to trigger this, both are with using [CEF](https://bitbucket.org/chromiumembedded/cef), although I suspect any "web view" environment where you have control over loading javascript at different points of the web view's lifetime could be used to trigger the situation described above.

In one Mac app I'm working on we are using React in the web views (which are not loaded over the internet — they are embedded in the app's executable) where each view needs to load React, Immutablejs, etc every time it "starts", which turns out to take quite a while when loaded from the file system of a computer with a mechanical disk.

So I figured we'd just load React once per Chromium render process by at app-compile time embed the various JS modules needed into the app's executable. (This is how we did it with nodejs and why nodejs can boot so quickly and still "load" a lot of JS code.)

First approach was to load react as a Chromium extension, and this kind of worked. This costs the least resources as Chromium/V8 interprets, loads and executes the extension only once per render process, so that any view being executed in that process gets access to the already-hot code. This approach mostly worked, but we'd see exceptions like the above and some other possibly related issues here and there. (Chromium extensions don't have access to console, window, etc and all code loaded is immutable.)

The second approach is to load React for every new JavaScript V8 context that Chromium is creating, meaning that the code is now mutable and has access to window, etc, but is loaded before any HTML document has been created.

The third approach and what turns out to be our best option right now, although slower than loading Rect as a Chromium extension, is to load reach on-demand from app memory though a custom URL scheme. Basically the loaded HTML document uses e.g. <script src="resource:react.js"> to have the app setup a byte stream that's "reading" from memory rather than disk or the network.

**It would be really great if React could be loaded as a Chromium extension.** Kind of how native dynamic libraries work (immutable, only provides code and constant data.)

**Comment by Aweary** [comment]
I'm not that familiar with CEF so hopefully someone from the React team can offer better or more detailed advice, but I would say that any code in the react-dom renderer is going to be dependent on having access to a `document`. Are you able to just wrap your `ReactDOM.render` call in a check that verifies the `document` is available? 

I suppose another option would be to implement a React renderer that wraps react-dom and only renders when the document is available.

**Comment by zpao** [comment]
Right now we do some work when the React is evaled which does some work based on if we think we're in a typical browser environment or not. This work is all guarded against with `ExecutionEnvironment.canUseDOM`, including the one you pointed out. Some of this doesn't happen at eval time but later during normal execution. It looks like your specific case is passing the environment check which was supposed to be "good enough" but isn't https://github.com/facebook/fbjs/blob/7c77c3c2ebcf0f1174512a66a08998dcff3d4cc5/src/core/ExecutionEnvironment.js#L14-L18) - we don't actually check for `document.documentElement` existing (and as pointed out there are a bunch of other places too).

Even if we did, as it stands, we don't support the changing of environment. We'd have to do some work to start handling it throughout.

**Comment by gaearon** [comment]
I think we’ll leave this as wontfix unless somebody motivated takes this and sends a PR that actually makes it work.
